File: 014/longest-common-prefix.go
package main

import "fmt"

func longestCommonPrefix(strs []string) string {
	if len(strs) == 0 {
		return ""
	}

	var loop func(int) int
	loop = func(i int) int {
		if i == len(strs[0]) {
			return i
		}

		c := strs[0][i]
		for _, str := range strs {
			if len(str) == i || str[i] != c {
				return i
			}
		}

		return loop(i + 1)
	}

	longestCommonPrefixLength := loop(0)
	return strs[0][0:longestCommonPrefixLength]
}

func main() {
	strs := []string{"flower", "flow", "flight"}
	fmt.Println(longestCommonPrefix(strs))
}

File: 014/longest-common-prefix.py
class Solution:
    def longestCommonPrefix(self, strs):
        """
        :type strs: List[str]
        :rtype: str
        """
        if len(strs) == 0:
            return ""
        
        def loop(i):
            if i == len(strs[0]):
                return i

            c = strs[0][i]
            for str in strs:
                if len(str) == i or str[i] != c:
                    return i

            return loop(i+1)

        longestCommonPrefixLength = loop(0)
        return strs[0][0:longestCommonPrefixLength]

strs = ["flower", "flow", "flight"]
print(Solution().longestCommonPrefix(strs))